core/metacling/src/TClingClassInfo.cxx

```diff
@@ -673,17 +673,26 @@ bool TClingClassInfo::HasDefaultConstructor() const
    // Note: This is could enhanced to also know about the ROOT ioctor
    // but this was not the case in CINT.
    //
-   if (!IsLoaded()) {
-      return false;
-   }
-   const CXXRecordDecl* CRD = llvm::dyn_cast<CXXRecordDecl>(GetDecl());
-   if (!CRD) {
-      // Namespaces do not have constructors.
-      return false;
+
+   using namespace ROOT::TMetaUtils;
+
+   if (!IsLoaded())
+      return EIOCtorCategory::kAbsent;
+
+   auto CRD = llvm::dyn_cast<CXXRecordDecl>(GetDecl());
+   // Namespaces do not have constructors.
+   if (!CRD)
+      return EIOCtorCategory::kAbsent;
+
+   const RConstructorType ioctortype_dflt("", *fInterp);
+   auto kind = CheckConstructor(CRD, ioctortype_dflt, *fInterp);
+
+   if (checkio && (kind == EIOCtorCategory::kAbsent)) {
+      const RConstructorType ioctortype_io("TRootIOCtor", *fInterp);
```

Comment:
See https://sft.its.cern.ch/jira/browse/ROOT-7723, we also support ```__void__&``` as an I/O constructor marker.
(see https://root.cern.ch/gitweb?p=root.git;a=commit;h=88b72f1c1bba34b22eab88f348f3a53980d494cc)

Which also points out that 'EIOCtorCategory' does not carry enough information.  (See and maybe use ROOT::TMetaUtils::HasIOConstructor)